### ISSUE TITLE
Support for negative sampling with implicit interaction data

### ISSUE THREAD

**Issue by opringle** [tensorflow/ranking#33]
I would like to train a model capable of recommending products to users. My training data is implicit, so each query-document pair has no relevance score.

For example:

```
1 uid:10 32:0.14 48:0.97  51:0.45
1 uid:10 1:0.15  31:0.75  32:0.24  49:0.6
1 uid:10 1:0.71  2:0.36   31:0.58  51:0.12
1 uid:4 1:0.15  31:0.75  32:0.24  49:0.6
1 uid:4 1:0.71  2:0.36   31:0.58  51:0.12
```

Here, the dataset contains 2 users. User "10" has 3 interactions and user "4" has 2 interactions. All interactions are simply 'click', without a relevance score. 

I want to randomly sample negatives (items a user did not click)  and train the model with a pairwise ranking loss.

[The paper](https://arxiv.org/pdf/1812.00073.pdf) states tf-ranking supports this, however, I cannot find any examples in the repo. Can someone point me in the right direction?

**Comment by bendersky** [comment]
Oliver,
You have two options here. 
1/ When you create the dataset also include items that were *not* clicked in the same session as the clicked items as negative examples. E.g., if user was shown items
X
Y
Z

and clicked Y, you can add X as negative (using the skip-above principle). You can also add both X and Z as negatives, but you'd might have to deal w/ some position bias effects.

2/ Randomly sample items from other queries and add them as "pseudo-negatives". E.g., in your case, you can add items from user 4 as negatives for user 10.

In general, 1/ is likely to be better, as it reflects true user and system behavior, and the negatives will have some semantic relation to the user. This is what we've done in the paper you cite. If you can't do 1/, 2/ is straightforward to implement as a pre-processing step, and then feed the resulting dataset into TF-Ranking.

Hope this helps!
Michael

**Comment by opringle** [comment]
This is very helpful Michael. 

Unfortunately (1) is not an option, since we have not stored what was shown to the user. I have one concern about approach 2:

 If I sample negatives at the time of creating the dataset, they will be fixed for the training procedure. In other papers, such as [neural collaborative filtering](https://www.comp.nus.edu.sg/~xiangnan/papers/ncf.pdf), the negatives are randomly sampled when creating the minibatch. This means the subset of negatives that are sampled can vary during training.

**Comment by bendersky** [comment]
Oliver,
Thank you for the pointer! We currently do not support in-batch negative sampling in TF-Ranking. We will consider adding this feature in the future releases, if there is more demand for it.
